### PR TITLE
 Removing redundant data from envelope

### PR DESCRIPTION
**Pull request by ssivov**

### CODE REVIEW

File: clarity.d.ts [hunk 1]
```diff
@@ -73,12 +73,16 @@ interface IPayload {
 }
 
 interface IEnvelope {
+  impressionId: string;
+  time?: number;
+  sequenceNumber?: number;
+}
+
+interface IClientInfo {
```
Comment (sarveshnagpal) [inline clarity.d.ts]: Have you considered keeping all this info in envelope except not repeating redundant stuff in subsequent envelopes to save on bytes. This will ensure that if for some reason top URL changes (like # appended) we will still capture it but won't waste bytes if nothing changed?

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "clarity-js",
-  "version": "0.1.16",
+  "version": "0.2.0",
```
Comment (sarveshnagpal) [inline package.json]: That's a jump... ambitious pull request :)

File: src/core.ts [hunk 3]
```diff
@@ -47,8 +48,12 @@ export function activate() {
     init();
   } catch (e) {
     onActivateErrorUnsafe(e);
+    return;
   }
 
+  // Send client information as the first event
+  sendClientInfo();
```
Comment (sarveshnagpal) [inline src/core.ts]: Based on my earlier comment, let's see if we can avoid adding another top-level JSON object. Envelope is (or was) supposed to address this.

File: package.json (hunk 2 above)
Comment (ssivov) [reply package.json]: I moved id up because there is a data format change

---
In reply to: [144364548](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144364548) [](ancestors = 144364548)

File: clarity.d.ts (hunk 1 above)
Comment (ssivov) [reply clarity.d.ts]: I have, but i thought that it wasn't clear enough. If we did that, we would be relying on previous payloads constantly and would have to re-scan entire data from start to end to see what was going on. Everything, except in rare cases url, is constant in this object. If we think it's important to track url changes, we should probably fire a separtate 'Redirect' event.

---
In reply to: [144364338](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144364338) [](ancestors = 144364338)

File: src/core.ts (hunk 3 above)
Comment (ssivov) [reply src/core.ts]: Please see my response to your previous comment

---
In reply to: [144364782](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144364782) [](ancestors = 144364782)

File: clarity.d.ts [hunk 4]
```diff
@@ -73,12 +73,16 @@ interface IPayload {
...
   clarityId: string;
   impressionId: string;
   url: string;
   version: string;
-  time?: number;
-  sequenceNumber?: number;
```
Comment (toby-walker) [inline clarity.d.ts]: Why are these nullable? I realize that is before this change but curious why.

File: src/core.ts [hunk 5]
```diff
@@ -58,9 +63,11 @@ export function activate() {
       state = State.Activated;
     } else {
       teardown();
+      return;
     }
   } catch (e) {
     onActivateError(e);
+    return;
```
Comment (toby-walker) [inline src/core.ts]: >return [](start = 4, length = 6)

What does adding these fix? Seems unrelated to format change.

File: src/core.ts [hunk 6]
```diff
@@ -358,6 +360,20 @@ function prepare() {
   return true;
 }
 
+function sendClientInfo() {
```
Comment (toby-walker) [inline src/core.ts]: >sendClientInfo [](start = 9, length = 14)

PageData seems better to me for a name.

File: clarity.d.ts (hunk 4 above)
Comment (ssivov) [reply clarity.d.ts]: All other properties in the envelope are constant throughout the impression, so we were storing an "abstract" envelope object without time and sequenceNo, which was later used in various places by appending relevant time and sequenceNo to it. 

This behavior is optimized now to avoid this pattern.

---
In reply to: [144390698](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144390698) [](ancestors = 144390698)

File: src/core.ts (hunk 5 above)
Comment (ssivov) [reply src/core.ts]: It is unrelated to the format change, but it was a small fix that I just corrected on the fly, while working on this PR :P

Essentially each of these catch blocks means that we've ran into a critical problem. These added 'return' statements make sure we terminate immediately after handling the problem

---
In reply to: [144390879](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144390879) [](ancestors = 144390879)

Comment (toby-walker) [reply src/core.ts]: I think this branch will *not* be merged for a while. Can you make these same fixes to the main (beta) version?

---
In reply to: [144425231](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144425231) [](ancestors = 144425231,144390879)

File: package.json (hunk 2 above)
Comment (toby-walker) [reply package.json]: I agree this is an appropriate bump. We will probably also go through multiple version on the "main" 0.1 series before this one goes out too. 

---
In reply to: [144371478](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144371478) [](ancestors = 144371478,144364548)

File: clarity.d.ts (hunk 1 above)
Comment (ssivov) [reply clarity.d.ts]: Synced offline: We renamed this to 'metadata' and we'd be passing it as a part of the first payload.

First payload will have the following structure:
{
  envelope: IEnvelope;
  events: IEvent[];
  metadata: IImpressionMetadata;
}

Subsequent payloads will just contain envelope and events.

---
In reply to: [144372367](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144372367) [](ancestors = 144372367,144364338)

File: src/core.ts (hunk 5 above)
Comment (ssivov) [reply src/core.ts]: Published a separate PR for that

---
In reply to: [144539040](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144539040) [](ancestors = 144539040,144425231,144390879)

File: src/core.ts (hunk 3 above)
Comment (ssivov) [reply src/core.ts]: Synced offline: We renamed this to 'metadata' and we'd be passing it as a part of the first payload.

First payload will have the following structure:
{
  envelope: IEnvelope;
  events: IEvent[];
  metadata: IImpressionMetadata;
}

Subsequent payloads will just contain envelope and events.

---
In reply to: [144372712](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144372712) [](ancestors = 144372712,144364782)

File: src/core.ts (hunk 6 above)
Comment (ssivov) [reply src/core.ts]: It also contains clientId - let's call it metadata

---
In reply to: [144391024](https://github.com/Microsoft/clarity-js/pull/39#discussion_r144391024) [](ancestors = 144391024)
